src/ucp.erl

```diff
@@ -67,13 +68,159 @@ cmd(C) when is_binary(C)    -> <<"unknown_",C/binary>>;
 cmd(C) when is_integer(C)   -> <<"unknown_",(integer_to_binary(C))/binary>>;
 cmd(C) -> <<"unknown_",(list_to_binary(io_lib:format("~p",[C])))/binary>>.
 
+-spec(decode(Str :: list() | binary()) -> {ok, map()} | {error, binary()}).
+decode(Str) when is_list(Str); is_binary(Str) ->
+    try
+        DecStr = parse(Str),
+        {ok, pl_to_map(DecStr, #{})}
+    catch
+       _:_ ->
+           io:format("Error: ~p~n", [erlang:get_stacktrace()]),
+           {error, <<"Invalid UCP Message">>}
+    end;
+decode(_) ->
+    {error, <<"Input to decode should be either binary or string">>}.
+
+-spec(pl_to_map(list(), map()) -> map()).
+pl_to_map([], AccMap) ->
+    AccMap;
+pl_to_map([{K, V}|R], AccMap) when is_list(V) ->
+    pl_to_map(R, AccMap#{atom_to_binary(K, utf8) => list_to_binary(V)});
+pl_to_map([{K, V}|R], AccMap) ->
+    pl_to_map(R, AccMap#{atom_to_binary(K, utf8) => V}).
+
+-spec(encode(PDU :: map()) -> {ok, binary()} | {error, binary()}).
+encode(PDU) when is_map(PDU) ->
+    [2|T_DATA] = pack(maps:fold(fun map_to_pl/3, [], PDU)),
+    [3|UCP_DATA] = lists:reverse(T_DATA),
+    {ok, list_to_binary(lists:reverse(UCP_DATA))};
+encode(_) ->
+    {error, <<"Input to encode should be map">>}.
+
+-spec(map_to_pl(term(), term(), list()) -> list()).
+map_to_pl(K, V, Acc) when is_binary(V) ->
+    [{binary_to_atom(K, utf8), binary_to_list(V)} | Acc];
```

Comment:
Rewrite with `maps:fold` directly

Reply:
I feel the current implementation makes it more readable..

Reply:
Yes the current implementation is fine... I meant to add to `maps:fold` it, instead of doing that outside
Instead of
```erlang
maps:fold(fun map_to_pl/3, [], PDU)
```
do
```erlang
map_to_pl(PDU) -> maps:fold(fun map_to_pl/3, [], PDU);
...
```

Reply:
👍